### PR TITLE
RavenDB-17097 Performance hint when index is outputing the source document

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17097 

### Additional description

When a user is including a source document in select output we will send a notification about that. 

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Config/Categories/PerformanceHintsConfiguration.cs [hunk 1]
```diff
@@ -58,5 +58,10 @@ public PerformanceHintsConfiguration()
         [DefaultValue(1024)]
         [ConfigurationEntry("PerformanceHints.Indexing.MaxNumberOfLoadsPerReference", ConfigurationEntryScope.ServerWideOrPerDatabase)]
         public int MaxNumberOfLoadsPerReference { get; set; }
+        
+        [Description("Alert when source document in indexed as field.")]
+        [DefaultValue(true)]
+        [ConfigurationEntry("PerformanceHints.Indexing.AlertWhenSourceDocumentIncludedAsProperty", ConfigurationEntryScope.ServerWideOrPerDatabase)]
```
Comment (ppekrol) [inline src/Raven.Server/Config/Categories/PerformanceHintsConfiguration.cs]: AlertWhenSourceDocumentIncludedAsProperty => AlertWhenSourceDocumentIncludedInOutput

Thoughts? @arekpalinski

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -4096,6 +4099,23 @@ public IndexingStatsAggregator GetLatestIndexingStat()
 
         public abstract void SaveLastState();
 
+        
+        protected void HandleSourceDocumentIncludedInMapOutput()
+        {
+            if (_alreadyNotifiedAboutIncludingDocumentInOutput || SourceDocumentIncludedInOutput == false || PerformanceHintsConfig.AlertWhenSourceDocumentIncludedAsProperty == false)
+                return;
+
+            _alreadyNotifiedAboutIncludingDocumentInOutput = true;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: maybe set it after the `Add`

File: test/SlowTests/Issues/RavenDB-17097.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        if (performanceHint.Item2 != null)
+            Assert.NotEqual($"Index '{index.IndexName}' is including the origin document in output.", performanceHint.Item2["Title"]);
+
+        WaitForUserToContinueTheTest(store);
+    }
+
+    [Fact]
+    public async Task FanOutIndexWithIncludedSourceDocumentWillSendNotification()
+    {
+        await NotificationTest<FanOutIndexWithSourceDocument>();
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-17097.cs]: you do not need async/await, can just return here

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Index.cs]: done

File: test/SlowTests/Issues/RavenDB-17097.cs (hunk 3 above)
Comment (maciejaszyk) [reply test/SlowTests/Issues/RavenDB-17097.cs]: done

File: src/Raven.Server/Config/Categories/PerformanceHintsConfiguration.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Raven.Server/Config/Categories/PerformanceHintsConfiguration.cs]: Done?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 4]
```diff
@@ -4096,6 +4099,23 @@ public IndexingStatsAggregator GetLatestIndexingStat()
...
+        {
+            if (_alreadyNotifiedAboutIncludingDocumentInOutput || SourceDocumentIncludedInOutput == false || PerformanceHintsConfig.AlertWhenSourceDocumentIncludedInOutput == false)
+                return;
+            
+            DocumentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                DocumentDatabase.Name,
+                $"Index '{Name}' is including the origin document in output.",
+                $"Please verify index definitions and consider a re-design of your entities or indexes for better indexing performance.",
+                PerformanceHintType.Indexing,
+                NotificationSeverity.Info,
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Index.cs]: I'd consider it to be pretty important so maybe let's put it as a warning to grab user's attention more

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Also I wonder if we shouldn't pass and check `numberOfOutputs` here. If this is a fanout index then it's even more obvious mistake and it's more dangerous so we can include that info into the notification.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 5]
```diff
@@ -4096,6 +4099,23 @@ public IndexingStatsAggregator GetLatestIndexingStat()
...
+        
+        protected void HandleSourceDocumentIncludedInMapOutput()
+        {
+            if (_alreadyNotifiedAboutIncludingDocumentInOutput || SourceDocumentIncludedInOutput == false || PerformanceHintsConfig.AlertWhenSourceDocumentIncludedInOutput == false)
+                return;
+            
+            DocumentDatabase.NotificationCenter.Add(PerformanceHint.Create(
+                DocumentDatabase.Name,
+                $"Index '{Name}' is including the origin document in output.",
+                $"Please verify index definitions and consider a re-design of your entities or indexes for better indexing performance.",
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Index.cs]: I wonder if we can express it better. What do you think about something like:

`Putting the whole document as one of the fields of the index entry isn't usually intentional. Please verify your index definition for better indexing performance`

?

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/AnonymousLuceneDocumentConverter.cs [hunk 6]
```diff
@@ -92,7 +93,7 @@ protected override int GetFields<T>(T instance, LazyStringValue key, LazyStringV
                     throw new InvalidOperationException($"Field '{property.Key}' is not defined. Available fields: {string.Join(", ", _fields.Keys)}.", e);
                 }
 
-                var numberOfCreatedFields = GetRegularFields(instance, field, value, indexContext, out var shouldSkip);
+                var numberOfCreatedFields = GetRegularFields(instance, field, value, indexContext, out var shouldSkip, sourceDocument);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/AnonymousLuceneDocumentConverter.cs]: Can we put the new parameter before `out` param? It's easier to notice the our parameters when they are put as the last ones

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Well, we're sending it only once per Index. So` numberOfOutputs` would be only valid for the first document.

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 5 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Yeah, this is better.

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/AnonymousLuceneDocumentConverter.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/AnonymousLuceneDocumentConverter.cs]: done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Yes, but still it's common that fanout indexes produce multiple entries on each document. Then we could underline that source document is outputted multiple times - which absolutely makes no sense

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**